raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -532,45 +655,59 @@ private OptionalLong maybeAppendAsLeader(LeaderState state, Records records) {
-                new InetSocketAddress(request.host(), request.port()),
-                request.bootTimestamp()
-            );
-            if (quorum.isVoter(request.replicaId())) {
-                connections.maybeUpdate(request.replicaId(), hostInfo);
-            }
+    /**
+     * Handle a FindQuorum request. Currently this API does not return any application errors.
+     */
+    private FindQuorumResponseData handleFindQuorumRequest(
+        RaftRequest.Inbound requestMetadata
+    ) {
+        FindQuorumRequestData request = (FindQuorumRequestData) requestMetadata.data;
+        HostInfo hostInfo = new HostInfo(
+            new InetSocketAddress(request.host(), request.port()),
+            request.bootTimestamp()
+        );
+        if (quorum.isVoter(request.replicaId())) {
+            connections.maybeUpdate(request.replicaId(), hostInfo);
         }
 
         List<FindQuorumResponseData.Voter> voters = new ArrayList<>();
         for (Map.Entry<Integer, Optional<HostInfo>> voterEntry : connections.allVoters().entrySet()) {
             FindQuorumResponseData.Voter voter = new FindQuorumResponseData.Voter();
             voter.setVoterId(voterEntry.getKey());
-            voterEntry.getValue().ifPresent(hostInfo -> {
-                voter.setHost(hostInfo.address.getHostString())
-                    .setPort(hostInfo.address.getPort())
-                    .setBootTimestamp(hostInfo.bootTimestamp);
+            voterEntry.getValue().ifPresent(voterHostInfo -> {
+                voter.setHost(voterHostInfo.address.getHostString())
+                    .setPort(voterHostInfo.address.getPort())
+                    .setBootTimestamp(voterHostInfo.bootTimestamp);
             });
             voters.add(voter);
         }
 
         return new FindQuorumResponseData()
-            .setErrorCode(error.code())
+            .setErrorCode(Errors.NONE.code())
             .setLeaderEpoch(quorum.epoch())
             .setLeaderId(quorum.leaderIdOrNil())
             .setVoters(voters);
     }
 
-    private void handleFindQuorumResponse(FindQuorumResponseData response) throws IOException {
-        // We only need to become a follower if the current leader is not elected
-        OptionalInt leaderIdOpt = optionalLeaderId(response.leaderId());
-        if (leaderIdOpt.isPresent() && leaderIdOpt.getAsInt() != quorum.localId) {
-            becomeFollower(leaderIdOpt.getAsInt(), response.leaderEpoch());
+    private boolean handleFindQuorumResponse(
+        RaftResponse.Inbound responseMetadata
+    ) throws IOException {
+        FindQuorumResponseData response = (FindQuorumResponseData) responseMetadata.data;
+        Errors error = Errors.forCode(response.errorCode());
+        OptionalInt responseLeaderId = optionalLeaderId(response.leaderId());
+        int responseEpoch = response.leaderEpoch();
+
+        // Always update voter connections if they are present, regardless of errors
+        updateVoterConnections(response);
```

Comment:
I seem to overlook this case, why does it make sense to always update connection information? Should we always look for response with equal or greater epoch?

A more general question is,  we do expect a broker to change their connection information through redeployment, which means the validity of connection information also goes down over time. Maybe we don't need to be strict in V1, but I feel in the long run only accepting up-to-date connection information seems less error-prone.

Reply:
The `bootTimestamp` is what we use to determine if a node's connection information is newer, so it's safe to use without considering epoch. The simulation tests fail without this check because we get stuck in states where one node that can't connect to other nodes and starts holding elections. However, the bumped epoch from these elections would prevent it from getting the latest information. Hence we update connections always.

I'm not super thrilled with the `bootTimestamp` proposal for what it's worth. I would prefer to have something which could guarantee monotonicity, but we also need something which can be derived from only local state. The best idea I've thought of is to store a separate epoch which is bumped on every boot. But that also is not ideal.

Reply:
I guess boot timestamp is good for now, although from our offline discussion it is not 100% reliable, but we could delay the solution until we discover more real issue under system tests.